### ISSUE TITLE
`@final` child `TypedDict` not a subtype of its parent?

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#2742]
I suppose this is a fallout of https://github.com/microsoft/pyright/issues/2608.

## Code & error
```py
from typing import TypedDict, final

class A(TypedDict):
    x: int
    y: str

@final
class B(A):
    z: list[int]

b: B = {"x": 1, "y": "foo", "z": [1, 2, 3]}
a: A = b
#      Expression of type "B" cannot be assigned to declared type "A"
#        "B" is incompatible with "A" because of a @final mismatch
```
Does this actually lead to some false negatives? It is really strange that a direct child of `A` is not a subtype of `A`.

**Comment by decorator-factory** [comment]
This pattern could be useful if you have some kind of "base" event, and then you have "leaves" with additional narrowing-down
```py
class BaseEvent(TypedDict):
    request_uuid: str
    common_field1: int
    common_field2: float
    ...

@final
class HatIsBought(BaseEvent):
    tag: Literal["hat_bought"]
    price: int

@final
class HatIsSold(BaseEvent):
    tag: Literal["hat_sold"]
    price: int
    client_name: str
    client_is_govt: bool
```

**Comment by erictraut** [comment]
Yes, this is a consequence of the #2608. You can either eliminate the `@final` and forego the ability to narrow unions in this case, or you can keep the `@final` and forego the ability to assign to the non-final base case. I don't think you can do both.
